src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdkAppEngineFlexibleStaging.java

```diff
@@ -60,47 +67,97 @@ public void stageFlexible(StageFlexibleConfiguration config) throws AppEngineExc
-          if (APP_ENGINE_CONFIG_FILES_WHITELIST.contains(configFile)) {
-            Files.copy(config.getAppEngineDirectory().toPath().resolve(configFile),
-                config.getStagingDirectory().toPath().resolve(configFile),
-                REPLACE_EXISTING);
-          } else if (configFile.equals("Dockerfile")) {
-            throw new AppEngineException("Found 'Dockerfile' in the App Engine directory."
-                + " Please move it to the Docker directory.");
-          } else {
-            throw new AppEngineException("Found an unexpected '" + configFile
-                + "' file in the App Engine directory.");
-          }
+  static void copyAppEngineContext(StageFlexibleConfiguration config, CopyService copyService)
+      throws IOException {
+    // Copy app.yaml and other App Engine config files to staging
+    String[] appEngineConfigFiles = config.getAppEngineDirectory().list();
+    if (appEngineConfigFiles != null) {
+      for (String configFile : appEngineConfigFiles) {
+        if (APP_ENGINE_CONFIG_FILES_WHITELIST.contains(configFile)) {
+          copyService.copyFileAndReplace(
+              config.getAppEngineDirectory().toPath().resolve(configFile),
+              config.getStagingDirectory().toPath().resolve(configFile));
+        } else if (configFile.equals("Dockerfile")) {
+          throw new AppEngineException("Found 'Dockerfile' in the App Engine directory."
+              + " Please move it to the Docker directory.");
+        } else {
+          throw new AppEngineException("Found an unexpected '" + configFile
+              + "' file in the App Engine directory.");
         }
       }
+    }
+  }
 
-      // Copy the JAR/WAR file to staging.
-      if (config.getArtifact() != null && config.getArtifact().exists()) {
-        Path destination = config.getStagingDirectory().toPath()
-            .resolve(config.getArtifact().toPath().getFileName());
-        Files.copy(config.getArtifact().toPath(), destination, REPLACE_EXISTING);
-      } else {
-        throw new AppEngineException("Artifact doesn't exist at '" + config.getArtifact().getPath()
-            + "'");
-      }
-    } catch (IOException e) {
-      throw new AppEngineException(e);
+  @VisibleForTesting
+  static void copyArtifact(StageFlexibleConfiguration config, CopyService copyService)
+      throws IOException {
+    // Copy the JAR/WAR file to staging.
+    if (config.getArtifact() != null && config.getArtifact().exists()) {
+      Path destination = config.getStagingDirectory().toPath()
+          .resolve(config.getArtifact().toPath().getFileName());
+      copyService.copyFileAndReplace(config.getArtifact().toPath(), destination);
+    } else {
+      throw new AppEngineException("Artifact doesn't exist at '" + config.getArtifact().getPath()
+          + "'");
+    }
+  }
+
+  @VisibleForTesting
+  static class CopyService {
+
+    void copyDirectory(Path src, Path dest) throws IOException {
```

Comment:
I'm instantiating the copyservice so I can mock it in the tests. I want the methods to be non-static.